lib/pacemaker/pcmk_sched_clone.c

```diff
@@ -249,77 +248,79 @@ pcmk__clone_apply_coloc_score(pe_resource_t *dependent,
                  colocation->id, dependent->id, primary->id,
                  pcmk_readable_score(colocation->score));
 
-    if (pcmk_is_set(primary->flags, pe_rsc_promotable)) {
-        if (colocation->primary_role == RSC_ROLE_UNKNOWN) {
-            // This isn't a role-specfic colocation, so handle normally
-            pe_rsc_trace(primary, "Handling %s as a clone colocation",
-                         colocation->id);
+    // Apply role-specific colocations
+    if (pcmk_is_set(primary->flags, pe_rsc_promotable)
+        && (colocation->primary_role != RSC_ROLE_UNKNOWN)) {
 
-        } else if (pcmk_is_set(dependent->flags, pe_rsc_provisional)) {
-            // We're placing the dependent
+        if (pcmk_is_set(dependent->flags, pe_rsc_provisional)) {
+            // We're assigning the dependent to a node
             pcmk__update_dependent_with_promotable(primary, dependent,
                                                    colocation);
-            return;
 
         } else if (colocation->dependent_role == RSC_ROLE_PROMOTED) {
-            // We're choosing roles for the dependent
+            // We're choosing a role for the dependent
             pcmk__update_promotable_dependent_priority(primary, dependent,
                                                        colocation);
-            return;
         }
+        return;
```

Comment:
Previously if we didn't match either condition (`pe_rsc_provisional` or `RSC_ROLE_PROMOTED`), we wouldn't return.

This might be correct though (do nothing if colocation is role-specific and we didn't match a condition).